src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,814 @@
+		state_cc(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex) = 1.e-5 / vol; // Injected tracer
+	});
+}
+
+void AddSupernova(amrex::MultiFab &mf, amrex::GpuArray<Real, AMREX_SPACEDIM> prob_lo, amrex::GpuArray<Real, AMREX_SPACEDIM> prob_hi,
+		  amrex::GpuArray<Real, AMREX_SPACEDIM> dx, SimulationData<NewProblem> const &userData, int level)
+{
+	// inject energy into cells with stochastic sampling
+	BL_PROFILE("QuokkaSimulation::Addsupernova()")
+
+	const Real cell_vol = AMREX_D_TERM(dx[0], *dx[1], *dx[2]); // cm^3
+	const Real rho_eint_blast = userData.E_blast / cell_vol;   // ergs cm^-3
+	const Real rho_blast = userData.M_ejecta / cell_vol;	   // g cm^-3
+	const int cum_sn = userData.SN_counter_cumulative;
+
+	const Real Lx = prob_hi[0] - prob_lo[0];
+	const Real Ly = prob_hi[1] - prob_lo[1];
+	const Real Lz = prob_hi[2] - prob_lo[2];
+
+	for (amrex::MFIter iter(mf); iter.isValid(); ++iter) {
+		const amrex::Box &box = iter.validbox();
+		auto const &state = mf.array(iter);
+		auto const &px = userData.blast_x->table();
+		auto const &py = userData.blast_y->table();
+		auto const &pz = userData.blast_z->table();
+		const int np = userData.nblast;
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			const Real xc = prob_lo[0] + static_cast<Real>(i) * dx[0] + 0.5 * dx[0];
+			const Real yc = prob_lo[1] + static_cast<Real>(j) * dx[1] + 0.5 * dx[1];
+			const Real zc = prob_lo[2] + static_cast<Real>(k) * dx[2] + 0.5 * dx[2];
+
+			for (int n = 0; n < np; ++n) {
+				Real x0 = NAN;
+				Real y0 = NAN;
+				Real z0 = NAN;
+				Real Rpds = 0.0;
+
+				x0 = std::abs(xc - px(n));
+				y0 = std::abs(yc - py(n));
+				z0 = std::abs(zc - pz(n));
+
+				if (x0 < 0.5 * dx[0] && y0 < 0.5 * dx[1] && z0 < 0.5 * dx[2]) {
+					state(i, j, k, HydroSystem<NewProblem>::density_index) += rho_blast;
+					state(i, j, k, HydroSystem<NewProblem>::energy_index) += rho_eint_blast;
+					state(i, j, k, HydroSystem<NewProblem>::internalEnergy_index) += rho_eint_blast;
+					state(i, j, k, Physics_Indices<NewProblem>::pscalarFirstIndex) += 1.e3 / cell_vol;
+
+					// printf("The total number of SN gone off=%d\n", cum_sn);
+					Rpds = 14. * std::pow(state(i, j, k, HydroSystem<NewProblem>::density_index) / Const_mH, -3. / 7.);
+					// printf("Rpds = %.2e pc\n", Rpds);
+				}
+			}
+		});
+	}
+}
+
+template <> void QuokkaSimulation<NewProblem>::computeBeforeTimestep()
+{
+	// compute how many (and where) SNe will go off on the this coarse timestep
+	// sample from Poisson distribution
+
+	const Real dt_coarse = dt_[0];
+	const Real domain_vol = geom[0].ProbSize();
+	const Real domain_area = geom[0].ProbLength(0) * geom[0].ProbLength(1);
+	const Real mean = 0.0;
+	const Real stddev = hscale / geom[0].ProbLength(2) / 2.;
+
+	const Real expectation_value = ks_sigma_sfr * domain_area * dt_coarse;
+
+	const int count = static_cast<int>(amrex::RandomPoisson(expectation_value));
```

Comment:
This is not necessary for now, but it would be useful to add a `// TODO` comment here to save (and restore) the RNG state in the `metadata.yaml` file in the future.